pkg/karmadactl/testing/helper.go

```diff
@@ -0,0 +1,152 @@
+package testing
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"encoding/pem"
+	"math/big"
+	"net"
+	"os"
+	"time"
+
+	v1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"sigs.k8s.io/yaml"
+)
+
+func generateMockKubeconfig() ([]byte, error) {
+	// set up our server certificate
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2022),
+		Subject: pkix.Name{
+			Organization:  []string{"foo"},
+			Country:       []string{"foo"},
+			Province:      []string{""},
+			Locality:      []string{"foo"},
+			StreetAddress: []string{"foo"},
+			PostalCode:    []string{"0000"},
+		},
+		IPAddresses:  []net.IP{net.IPv4(127, 0, 0, 1), net.IPv6loopback},
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(10, 0, 0),
+		SubjectKeyId: []byte{1, 2, 3, 4, 6},
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:     x509.KeyUsageDigitalSignature,
+	}
+
+	// create private and public key
+	caPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the CA
+	caBytes, err := x509.CreateCertificate(rand.Reader, cert, cert, &caPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return nil, err
+	}
+
+	certPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, err
+	}
+
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, cert, &certPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return nil, err
+	}
+
+	caPEM := new(bytes.Buffer)
+	_ = pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caBytes,
+	})
+
+	certPEM := new(bytes.Buffer)
+	_ = pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	certPrivKeyPEM := new(bytes.Buffer)
+	_ = pem.Encode(certPrivKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certPrivKey),
+	})
+
+	config := v1.Config{
+		Kind:           "Config",
+		APIVersion:     "v1",
+		Preferences:    v1.Preferences{},
+		CurrentContext: "context-c1",
+		Clusters: []v1.NamedCluster{
+			{
+				Name: "c1",
+				Cluster: v1.Cluster{
+					CertificateAuthorityData: caPEM.Bytes(),
+					Server:                   "http://127.0.0.1:5443",
+				},
+			},
+		},
+		Contexts: []v1.NamedContext{
+			{
+				Name: "context-c1",
+				Context: v1.Context{
+					Cluster:  "c1",
+					AuthInfo: "user1",
+				},
+			},
+		},
+		AuthInfos: []v1.NamedAuthInfo{
+			{
+				Name: "user1",
+				AuthInfo: v1.AuthInfo{
+					ClientCertificateData: certPEM.Bytes(),
+					ClientKeyData:         certPrivKeyPEM.Bytes(),
+				},
+			},
+		},
+	}
+
+	jsonData, err := json.Marshal(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return yaml.JSONToYAML(jsonData)
+}
+
+// CreateMockKubeconfigInTmpDir create fake kubeconfig file in tmp/kubeconfig.yaml
+func CreateMockKubeconfigInTmpDir() error {
```

Comment:
In order to prevent `f` from being closed when it fails to be written to `f`, is it better?

```diff
--- a/pkg/karmadactl/testing/helper.go
+++ b/pkg/karmadactl/testing/helper.go
@@ -139,11 +139,13 @@ func CreateMockKubeconfigInTmpDir() error {
        if err != nil {
                return err
        }
+       defer f.Close()
+
        if _, err := f.Write(yamlData); err != nil {
                return err
        }

-       return f.Close()
+       return nil
 }
```

Reply:
Why do we need this function?

Reply:
To generate a fake kubeconfig file, in pkg/karmadactl, there are many functions need one kubeconfig file to test.